orm/bucket.go

```diff
@@ -189,7 +189,15 @@ func (b bucket) Parse(key, value []byte) (Object, error) {
 	return obj, nil
 }
 
-// Save will write a model, it must be of the same type as proto
+// Save will write a model, it must be of the same type as bucket.proto.
+// If model.Value() has GetMetadata(), then we set the LastModified height
+// to current block height.
+//
+// TODO: how do we pass this in?
+// 1. Set the height in weave.KVStore when we cachewrap?
+// 2. Pass the height explicitly in Save?
+// 3. Pass weave.Context to Save (and other methods)?
+// 4. Other ideas?
```

Comment:
This is a very good question. Without solving this problem, this whole change is not that great.

I think if we will decide to implement it then all buckets must use it. It cannot be optional.
I think as you suggested maintaining this value must be transparent to the user to avoid mistakes (wrong value, forgot to set etc). With the current `Save` method notation, I believe this is not possible. There are two ways I can think of to make it possible.
1. Introduce another argument to the `Save` method. Context or some other value. Pprovided variable cannot be block height directly. Using `context.Context` while in my opinion the best of all options is probably a missuse.
2. Create an `UnboundBucket` that will be assigned to every handler structure and used everywhere. To use it, one must call `b := unboundBucket.Bind(blockHeight)`.

What about places like genesis, where there is no block height?

Reply:
Yeah... big API changes.

> Create an UnboundBucket that will be assigned to every handler structure and used everywhere. To use it, one must call b := unboundBucket.Bind(blockHeight).

I kind of like this idea... In fact `UnboundBucket` is like the constructor....

```go
type UnboundBucket interface {
  Bind(info BindInfo) Bucket
  WithIndex()
  WithMultiIndex()
  // ...
}

type BindInfo struct {
  int64 Height
  // other ???
}
```

All initializer functions work on `UnboundBucket`

Then Bucket is just read/write queries, but not the setup functions.

This ties into other API changes as well, such as #645 and warrants a larger discussion

Reply:
> Introduce another argument to the Save method. Context or some other value. Pprovided variable cannot be block height directly. Using context.Context while in my opinion the best of all options is probably a missuse.

Yes, this is the easiest and my first thought. But I also felt it was a mis-use. I assume the Bucket is "dumb" and driven by the handler. Unless we want to put lots of app/auth-specific logic in the Bucket, I would avoid `weave.Context`. However, I think this is another approach, and kind of takes the start of the migration.Bucket and extends it to add even more info. We could even add Bucket based authentication controls and such. But.... I think that probably makes the whole design less maintainable and extensible

Reply:
To be honest I think we could keep it simple and use context to pass these variables as we are already using it for handlers and this is consistent.
If you want it to be cleaner - we can add another param to ‘bucket.Save’ which could be a struct with all future params. And we can do ‘ToBucketParams(weave.Context)’ for extracting the info.

Reply:
We could even introduce a weave.Context method for that to reduce verbosity. Or just pass the context directly and get what we need from it while handling all the errors. 

If this approach does not work at some point in future - we could always go for a larger change later.

Reply:
And you recommend, just adding this Context/BucketParams to Save?
Or to all methods?

This goes into 0.17 for sure, not 0.16.1

Reply:
This is a really good question.

One solution to which would be to enhance the bucket with context as a decorator, so that it'd have the context as part of bucket struct. Then you always have fresh context, same api as before in every extension and the ability to operate on it.

Reply:
> One solution to which would be to enhance the bucket with context as a decorator, so that it'd have the context as part of bucket struct. Then you always have fresh context, same api as before in every extension and the ability to operate on it.

Which is the same idea as @husio with `unboundBucket.Bind()`...

Reply:
`context.Context` is so much more than scoped values. Not only other functionalities of context would not be utilized by the bucket, but they would be ignored.
Instead of `context.Context` we could create a structure `weave.BlockState` and pass this around. All attributes are typed and there is no need for second-guessing during the runtime if a variable is attached and of the right type.

That brings us to a discussion about the domain boundaries. What should bucket "know"? Should it understand what a handler is? Should it be aware of the synchronization mechanism (tendermint) we are using?

Reply:
> Which is the same idea as @husio with `unboundBucket.Bind()`...

The idea is similar, but I'm not sure there's need to introduce new interfaces for that. We could enhance what we have to work like this and make seamless integration. 

> That brings us to a discussion about the domain boundaries. What should a bucket "know"? Should it understand what a handler is? Should it be aware of the synchronization mechanism (tendermint) we are using?

I'm not sure we have to go and overthink this. At the moment we need to pass a couple fields related to current block height/time. If we pass that as decorator, I don't see how it matters whether we pass context or, say, a structure extracted from such context. Either is easy to do and easy to refactor should the need arise as there should Ideally be only one place where all the buckets are bound, similar to Initializer interface.

Reply:
The need for context was to allow modules to add their own values dynamically, for permissioning. The actual block data can be hard-coded in the framework. And maybe that is a cleaner idea, even if it breaks the handler api, at least we know what we have and we just use context for Auth state, or other module specific handlers.